tools/pubsys/src/aws/validate_ami/mod.rs

```diff
@@ -0,0 +1,799 @@
+    for mut image in expected_images {
+        // If the image is expected to be public, the specific launch permissions are irrelevant
+        // and were not retrieved for the actual image
+        if image.public {
+            image.launch_permissions = None;
+        }
+        results.insert(AmiValidationResult::new(
+            image.id.clone(),
+            image.clone(),
+            actual_images.get(&image.id).map(|v| v.to_owned()),
+            region.clone(),
+        ));
+    }
+
+    results
+}
+
+type RegionName = String;
+type AmiId = String;
+
+/// Parse the file holding image values. Return a HashMap of Region mapped to a vec of ImageDefs
+/// for that region.
+pub(crate) async fn parse_expected_amis(
+    expected_amis_path: &PathBuf,
+) -> Result<HashMap<Region, Vec<ImageDef>>> {
+    // Parse the JSON file as a HashMap of region_name, mapped to a JSON value
+    let expected_amis: HashMap<RegionName, serde_json::Value> = serde_json::from_reader(
+        &File::open(expected_amis_path.clone()).context(error::ReadExpectedImagesFileSnafu {
+            path: expected_amis_path,
+        })?,
+    )
+    .context(error::ParseExpectedImagesFileSnafu)?;
+
+    // If the value for each Region is a JSON object instead of an array, wrap it in an array
+    let vectored_images = expected_amis
+        .into_iter()
+        .map(|(region, value)| {
+            let image_values = match value.as_array() {
+                Some(array) => array.to_owned(),
+                None => vec![value],
+            };
+            (Region::new(region), image_values)
+        })
+        .collect::<HashMap<Region, Vec<serde_json::Value>>>();
+
+    // Parse the arrays of JSON objects into vecs of ImageDefs
+    let image_map = vectored_images
+        .into_iter()
+        .map(|(region, image_values)| {
+            let images = image_values
+                .into_iter()
+                .map(serde_json::from_value::<ImageDef>)
+                .collect::<serde_json::Result<Vec<ImageDef>>>();
+            match images {
+                Ok(images) => Ok((region, images)),
+                Err(e) => Err(e),
+            }
+        })
+        .collect::<serde_json::Result<HashMap<Region, Vec<ImageDef>>>>()
+        .context(error::ParseExpectedImagesFileSnafu)?;
+
+    Ok(image_map)
+}
+
+/// Common entrypoint from main()
+pub(crate) async fn run(args: &Args, validate_ami_args: &ValidateAmiArgs) -> Result<()> {
+    let results = validate(args, validate_ami_args).await?;
+
+    if validate_ami_args.json {
+        println!(
+            "{}",
+            serde_json::to_string_pretty(&results.get_json_summary())
+                .context(error::SerializeResultsSummarySnafu)?
+        )
+    } else {
+        println!("{}", results);
+    }
+    Ok(())
+}
+
+mod error {
+    use snafu::Snafu;
+    use std::path::PathBuf;
+
+    #[derive(Debug, Snafu)]
+    #[snafu(visibility(pub(super)))]
+    pub(crate) enum Error {
```

Comment:
Please check the error enum variants and ensure they're being used. I see a few that aren't being used anywhere. e.g. `ReadValidationConfig`, `ParseValidationConfig`, `MissingField` etc.

Reply:
> Please check the error enum variants and ensure they're being used. I see a few that aren't being used anywhere. e.g. ReadValidationConfig, ParseValidationConfig, MissingField etc.

As an aside, the reason you have to check is because the code that Snafu generates "uses" each enum variant so the compiler can't warn you about unused variants.